Add support for data versioning

tessellator opened the pull request:
This pull request adds support for `data-version` and `data-default` bindings, which allow the designer to defer the decision of which version of markup to render to the backend. These are similar to bindings of the same names in Pakyow, though there are some minor differences.

As a silly little example of how these might be used, let's consider two different layouts for a blog post.

``` html
<div data-scope="post">
  <h1 data-prop="title">Post Title</h1>

  <p data-default data-prop="full-text">Full Post Text</p>
  <p data-version="short" data-prop="short-text">Short Post Text</p>
</div>
```

The value of the `data-version` attribute is the name of the version. It can have multiple values (e.g., `data-version="short medium"`). Note that the `h1` element does not have a version assigned, and so it appears in every version.

To specify a version to use, support for a `:version` option has been added to the default implementation of `glimpse.views/bind`. The value can be a string or keyword that name the desired version, `nil` to indicate the default version should be used, or a function that accepts the binding data and returns a string, keyword, or `nil`.

To bind with a function, the following code might be used:

``` clojure
(glimpse.views/bind view
                    :post
                    [{:title "Full" :full-text "Some Full Text"}
                     {:title "Short" :short-text "Some Short Text" :version :short}]
                    :version #(:version %))
```

For the first piece of data, the version function returns `nil`, so the `data-default` element is used. In the second case, the function returns `short`, so the `data-version="short"` element is used.

The `:version` option on bind only works at the top level; it does not reapply the version selector to sub-scopes. The appropriate method to solve this problem is left to a future issue.

Fixes #2.

tessellator committed:
Add support for data versioning.

This commit adds the data-version and data-default bindings.

Fixes #2.

Signed-off-by: Chad Taylor <taylor.thomas.c@gmail.com>

tessellator committed:
Return version name from fn only when non-nil.

tessellator committed:
Enable :version option to bind.

hby commented:
First thought: It would be nice if the `:version` option took a sequence so one could do something like

``` clojure
...
:version (cycle :even-row :odd-row))
```

tessellator commented:
That's a good idea. I'll look into it.

tessellator committed:
Add support for binding with version collections.

This commit adds support for binding with collections of versions on a
single bind-scope command (e.g., `:version (cycle :even :odd)`). It also
adds support for embedding version data using the
::glimpse.views/version key.

tessellator committed:
Fix keyword namespacing issue.

tessellator commented:
@hby Please check out these changes. It adds support for providing a sequence option to `:version` as well as providing a mechanism to embed the version information in the data itself via a `:glimpse.views/version` key. This enables recursive version selection.
